From 09afe9c37a9fee57a6fe1a7b2d2d9dc8ce82d6ec Mon Sep 17 00:00:00 2001
From: xzhang <xzhang@xzdt>
Date: Tue, 28 Jul 2015 11:39:37 -0700
Subject: [PATCH 0208/1164] CDH-29924: HIVE-11288: Avro SerDe InstanceCache
 returns incorrect schema (Greg via Xuefu)

Conflicts:
	serde/src/java/org/apache/hadoop/hive/serde2/avro/InstanceCache.java
	serde/src/test/org/apache/hadoop/hive/serde2/avro/TestInstanceCache.java
---
 .../hadoop/hive/serde2/avro/InstanceCache.java     |   10 ++++++----
 .../hadoop/hive/serde2/avro/TestInstanceCache.java |   16 +++++++++++++++-
 2 files changed, 21 insertions(+), 5 deletions(-)

diff --git a/serde/src/java/org/apache/hadoop/hive/serde2/avro/InstanceCache.java b/serde/src/java/org/apache/hadoop/hive/serde2/avro/InstanceCache.java
index d848005..9c73875 100644
--- a/serde/src/java/org/apache/hadoop/hive/serde2/avro/InstanceCache.java
+++ b/serde/src/java/org/apache/hadoop/hive/serde2/avro/InstanceCache.java
@@ -21,6 +21,8 @@
 import org.apache.commons.logging.LogFactory;
 
 import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
 
 /**
  * Cache for objects whose creation only depends on some other set of objects
@@ -32,7 +34,7 @@
  */
 public abstract class InstanceCache<SeedObject, Instance> {
   private static final Log LOG = LogFactory.getLog(InstanceCache.class);
-  HashMap<Integer, Instance> cache = new HashMap<Integer, Instance>();
+  Map<SeedObject, Instance> cache = new HashMap<SeedObject, Instance>();
   
   public InstanceCache() {}
 
@@ -43,15 +45,15 @@ public InstanceCache() {}
   public Instance retrieve(SeedObject hv) throws AvroSerdeException {
     if(LOG.isDebugEnabled()) LOG.debug("Checking for hv: " + hv.toString());
 
-    if(cache.containsKey(hv.hashCode())) {
+    if(cache.containsKey(hv)) {
       if(LOG.isDebugEnabled()) LOG.debug("Returning cache result.");
-      return cache.get(hv.hashCode());
+      return cache.get(hv);
     }
 
     if(LOG.isDebugEnabled()) LOG.debug("Creating new instance and storing in cache");
 
     Instance instance = makeInstance(hv);
-    cache.put(hv.hashCode(), instance);
+    cache.put(hv, instance);
     return instance;
   }
 
diff --git a/serde/src/test/org/apache/hadoop/hive/serde2/avro/TestInstanceCache.java b/serde/src/test/org/apache/hadoop/hive/serde2/avro/TestInstanceCache.java
index 1df88ee..232f6355ab 100644
--- a/serde/src/test/org/apache/hadoop/hive/serde2/avro/TestInstanceCache.java
+++ b/serde/src/test/org/apache/hadoop/hive/serde2/avro/TestInstanceCache.java
@@ -23,9 +23,23 @@
 
 public class TestInstanceCache {
   private static class Foo {
+
+    private int value = 42;
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (o == null || getClass() != o.getClass()) return false;
+
+      Foo foo = (Foo) o;
+
+      return value == foo.value;
+
+    }
+
     @Override
     public int hashCode() {
-      return 42;
+      return value;
     }
   }
 
-- 
1.7.9.5

